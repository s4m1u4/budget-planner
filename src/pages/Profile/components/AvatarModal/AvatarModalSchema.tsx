import * as yup from "yup";

export const AvatarModalSchema = yup.object({
  avatar: yup.string().required("Avatar is required"),
});
