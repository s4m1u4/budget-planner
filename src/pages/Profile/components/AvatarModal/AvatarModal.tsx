import React, { FC } from "react";
import { useFormik } from "formik";
import {
  ButtonComponent,
  InputComponent,
  ModalComponent,
} from "components/shared";
import { AvatarModalSchema } from "./AvatarModalSchema";
import { IAvatarLink, IUserData } from "types";

import { ButtonGroup, Title } from "./AvatarModal.styles";

interface AvatarModalProps {
  open: boolean;
  handleClose: () => void;
  setNewAvatar: (avatar: IAvatarLink) => void;
  getUserData: () => IUserData;
  setUserData: () => void;
  onSubmit: (value: IAvatarLink) => void;
}

export const AvatarModal: FC<AvatarModalProps> = ({
  open,
  handleClose,
  setNewAvatar,
  getUserData,
  setUserData,
  onSubmit,
}) => {
  const handleSubmit = async (values: IAvatarLink) => {
    await setNewAvatar(values);
    await getUserData();
    handleClose();
    await setUserData();
    onSubmit(values);
    formik.resetForm();
  };

  const formik = useFormik({
    initialValues: {
      avatar: "",
    },
    validationSchema: AvatarModalSchema,
    onSubmit: handleSubmit,
  });

  return (
    <ModalComponent open={open} onClose={handleClose}>
      <Title>Set a new avatar</Title>
      <form onSubmit={formik.handleSubmit}>
        <InputComponent
          fullWidth
          label="Link to avatar"
          type="text"
          name="avatar"
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          value={formik.values.avatar || ""}
          error={(formik.touched.avatar && !!formik.errors.avatar) || false}
          helperText={(formik.touched.avatar && formik.errors.avatar) || ""}
        />
        <ButtonGroup>
          <ButtonComponent
            type="submit"
            color="success"
            onClick={formik.handleSubmit}
          >
            Change
          </ButtonComponent>
          <ButtonComponent type="button" color="error" onClick={handleClose}>
            Cancel
          </ButtonComponent>
        </ButtonGroup>
      </form>
    </ModalComponent>
  );
};
