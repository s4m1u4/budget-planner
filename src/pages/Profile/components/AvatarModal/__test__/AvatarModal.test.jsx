import { AvatarModal } from "../AvatarModal";
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const renderComponent = () => {
  return render(
    <AvatarModal
      open={open}
      setNewAvatar={setNewAvatar}
      handleClose={handleClose}
      onSubmit={onSubmit}
      getUserData={getUserData}
      setUserData={setUserData}
    />
  );
};

const renderComponentAsync = async () => {
  await act(
    async () =>
      await render(
        <AvatarModal
          open={open}
          setNewAvatar={setNewAvatar}
          handleClose={handleClose}
          onSubmit={onSubmit}
          getUserData={getUserData}
          setUserData={setUserData}
        />
      )
  );
};

const open = true;
const onSubmit = jest.fn();
const handleClose = jest.fn();
const setUserData = jest.fn();
const getUserData = jest.fn();
const setNewAvatar = jest.fn();

describe("Avatar modal component", () => {
  it("snapshot matching", () => {
    const { baseElement } = renderComponent();
    expect(baseElement).toMatchSnapshot();
  });

  beforeEach(() => {
    renderComponent();
  });

  describe("should render", () => {
    it("a title", () => {
      expect(screen.getByText(/Set a new avatar/i)).toBeInTheDocument();
    });

    it("an input with name='avatar'", () => {
      expect(screen.getByRole("textbox")).toHaveAttribute("name", "avatar");
    });

    it('a button "Change', () => {
      expect(
        screen.getByRole("button", { name: /Change/i })
      ).toBeInTheDocument();
    });

    it('a button "Cancel', () => {
      expect(
        screen.getByRole("button", { name: /Cancel/i })
      ).toBeInTheDocument();
    });

    it("error message when input value is empty and focus disappeared", async () => {
      expect(screen.queryByText(/Avatar is requred/i)).toBeNull();
      fireEvent.focus(screen.getByRole("textbox"));
      fireEvent.blur(screen.getByRole("textbox"));
      await renderComponentAsync();
      expect(screen.getByText(/Avatar is required/i)).toBeInTheDocument();
    });

    it("error message when submitting a form with empty input", async () => {
      expect(screen.queryByText(/Avatar is requred/i)).toBeNull();
      fireEvent.click(screen.getByRole("button", { name: /Change/i }));
      await renderComponentAsync();
      expect(screen.getByText(/Avatar is required/i)).toBeInTheDocument();
    });
  });

  it("successful form submission", async () => {
    userEvent.type(
      screen.getByRole("textbox"),
      "https://cdn.dribbble.com/users/3293507/screenshots/14667603/media/d8cbe035a61f64afdf6deabca5182842.jpg?compress=1&resize=400x300"
    );

    userEvent.click(screen.getByRole("button", { name: /Change/i }));

    await waitFor(() =>
      expect(onSubmit).toHaveBeenCalledWith({
        avatar:
          "https://cdn.dribbble.com/users/3293507/screenshots/14667603/media/d8cbe035a61f64afdf6deabca5182842.jpg?compress=1&resize=400x300",
      })
    );

    expect(onSubmit).toHaveBeenCalledTimes(1);
    expect(screen.getByRole("textbox")).toHaveValue("");
  });
});
