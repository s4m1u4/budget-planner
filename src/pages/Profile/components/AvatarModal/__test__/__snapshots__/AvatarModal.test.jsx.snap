// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar modal component snapshot matching 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    aria-hidden="true"
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiBox-root css-5fyqun"
      tabindex="-1"
    >
      <h3
        class="css-ngqp4j"
      >
        Set a new avatar
      </h3>
      <form>
        <div
          class="css-acwcvw"
        >
          <label
            class="css-zkehil"
          >
            Link to avatar
          </label>
          <input
            class="css-1pabijg"
            name="avatar"
            type="text"
            value=""
          />
          <p
            class="css-1ghz2lr"
          />
        </div>
        <div
          class="css-awol2u"
        >
          <button
            class="css-1p4hstr"
            color="success"
            type="submit"
          >
            Change
          </button>
          <button
            class="css-egajlx"
            color="error"
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiBox-root css-5fyqun"
      tabindex="-1"
    >
      <h3
        class="css-ngqp4j"
      >
        Set a new avatar
      </h3>
      <form>
        <div
          class="css-acwcvw"
        >
          <label
            class="css-zkehil"
          >
            Link to avatar
          </label>
          <input
            class="css-1pabijg"
            name="avatar"
            type="text"
            value=""
          />
          <p
            class="css-1ghz2lr"
          />
        </div>
        <div
          class="css-awol2u"
        >
          <button
            class="css-1p4hstr"
            color="success"
            type="submit"
          >
            Change
          </button>
          <button
            class="css-egajlx"
            color="error"
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
